pilopt/src/inliner.rs

```diff
@@ -89,17 +168,45 @@ fn find_inlinable_variables<T: FieldElement, V: Ord + Clone + Hash + Eq>(
 
 /// Checks whether a substitution is valid under `max_degree` constraint.
 fn is_valid_substitution<T: FieldElement, V: Ord + Clone + Hash + Eq>(
+    var_index: &VarIndex<V>,
     var: &V,
     expr: &QuadraticSymbolicExpression<T, V>,
     constraint_system: &ConstraintSystem<T, V>,
     max_degree: usize,
+    // these are already marked for removal, so we can ignore them
+    ignore_constraints: &[usize],
 ) -> bool {
     let replacement_deg = qse_degree(expr);
 
-    !constraint_system
-        .expressions()
-        .map(|constraint| qse_degree_with_virtual_substitution(constraint, var, replacement_deg))
-        .any(|deg| deg > max_degree)
+    let is_valid_constraints = constraint_system
+        .algebraic_constraints
+        .iter()
+        .enumerate()
+        .filter(|(idx, _)| {
+            var_index.constraints[var].contains(idx) && !ignore_constraints.contains(idx)
```

Comment:
Why not iter over `var_index.constaints[var]` directly? Because it does not store the index?